plugins/radar-android-google-places/src/main/java/org.radarbase.passive.google.places/GooglePlacesManager.kt

```diff
@@ -0,0 +1,456 @@
+                    client.fetchPlace(detailsPlaceRequest)
+                            .addOnSuccessListener { placeInfo: FetchPlaceResponse ->
+                                val addressComponents = placeInfo.place.addressComponents?.asList()
+                                val city = findComponent(addressComponents, CITY_KEY)
+                                val state = findComponent(addressComponents, STATE_KEY)
+                                val country = findComponent(addressComponents, COUNTRY_KEY)
+                                send(placesInfoTopic, GooglePlacesInfo(currentTime, currentTime, types, city, state, country, it, placeLikelihood, fromBroadcastRegistration))
+                                logger.info("Google Places data with additional info sent")
+                            }
+                            .addOnFailureListener { ex ->
+                                when (ex.javaClass) {
+                                    ApiException::class.java -> {
+                                        val exception = ex as ApiException
+                                        handleApiException(exception, exception.statusCode)
+                                    }
+                                    else -> {
+                                        send(placesInfoTopic, GooglePlacesInfo(currentTime, currentTime, types, null, null, null, placeId, placeLikelihood, fromBroadcastRegistration))
+                                        logger.info("Google Places data sent")
+                                    }
+                                }
+                            }
+                }
+            }
+        }
+    }
+
+    private fun findComponent(addressComponents: MutableList<AddressComponent>?, component: String): String? =
+             addressComponents?.firstOrNull { address ->
+                 address.types.any { it.equals(component) }
+             }?.name
+
+    private fun handleApiException(exception: ApiException, statusCode: Int) {
+        when (statusCode) {
+            INVALID_API_CODE -> {
+                logger.error("Invalid Api key, disconnecting now")
+                onDisconnect()
+            }
+            LOCATION_UNAVAILABLE_CODE -> logger.warn("Location not enabled yet")
+            else -> logger.error("ApiException occurred with status code $statusCode", exception)
+        }
+    }
+
+    private fun onDisconnect() {
+        if (!isClosed) {
+            numOfAttempts = preferences.getInt(NUMBER_OF_ATTEMPTS_KEY, 0)
+            val currentDelay = (baseDelay + 2.0.pow(numOfAttempts)).toLong()
+            placeHandler.delay(currentDelay * 1000) {
+                if (currentDelay < maxDelay) {
+                    numOfAttempts++
+                    preferences
+                        .edit()
+                        .putInt(NUMBER_OF_ATTEMPTS_KEY, numOfAttempts).apply()
+                }
+                    disconnect()
+            }
+        }
```

Comment:
IIRC, onDisconnect is only called if `isClosed` is false, and no need to call `disconnect` again.
So the if statement could be removed, and the disconnect call as well

Reply:
May I inquire whether `onDisconnect` is present within the superclasses or if it has been specified elsewhere? Could you kindly provide clarification regarding the precise location of the onDisconnect method call? Regrettably, I am unable to locate it within the superclasses.

Reply:
Ah, I'm sorry, it should be `onClose()`

Reply:
Okay, Thanks.